File: src/model/mongo/patient.model.js
/**
 * Definición del esquema de la colección 'patients' para la base de datos MongoDB.
 * Este esquema describe la estructura de los documentos de pacientes, incluyendo campos como nombre, apellido, DNI, etc.
 */
import mongoose from "mongoose";
import mongoosePaginate from 'mongoose-paginate-v2'

// Verificamos si el modelo está en la caché para evitar errores
/* if (mongoose.models["patients"]) {
  delete mongoose.models["patients"];
} */

//Esquema de dientes
const toothSchema = new mongoose.Schema({
  tooth: { type: Number, required: true },
  caries: {
    vestibular: { type: String, default: "0" },
    mesial: { type: String, default: "0" },
    oclusal: { type: String, default: "0" },
    distal: { type: String, default: "0" },
    lingual: { type: String, default: "0" }
  },
  corona: { type: Boolean, default: false },
  extracted: { type: Boolean, default: false },
  allcaries: { type: Boolean, default: false },
  incurable: { type: Boolean, default: false },
  malposition: { type: Boolean, default: false },
  periodontal: { type: Boolean, default: false },
  inscrustration: { type: Boolean, default: false }
}, { _id: false }); // _id: false evita que cada diente tenga su propio _id

const patientsSchema = new mongoose.Schema({
    name: {
        type: String,
        required: true
    },
    lastName:{
        type: String,
        required: true
    },
    birth:{
        type: String,
        required: true
    },
    typeDNI:{
        type: String,
    },
    dni:{
        type: String,
        unique: true,
        required: true
    },
    sex:{
        type: String,
        required: true
    },
    address:{
        type: String,
        required: true
    },
    phone:{
        type: String,
        required: true
    },
    email:{
        type: String,
        required: true
    },
    medicalCoverage:{
        type: String,
        required: true
    },
    nAffiliate:{
        type: String,
        required: true
    },
    treatments: [
        {
            name:{
                type: String,
                required: true 
            },
            dateStart: {
                type: String,
                required: true,
            },
            dateEnd: {
                type: String,
                required: true,
            },
            status: {
                type: String,
                enum:["progress","pending", "finalized", "canceled" ]
            },
            IDPatient:{
                type: String,
            },
            IDdoctor: {
                type: String,
            }
        }
    ],
    observations: [
        {
            name:{
                type: String,
                required: true 
            },
            date: {
                type: String,
                required: true,
            },
            status: {
                type: String,
                enum:["pending", "finalized", "canceled" ]  
            }  
        }
    ],
    dentalStatus: [toothSchema],
    status:{
        type: String,
        enum: ["active", "inactive", "blocked"],
        default: "active"
    },
    idDoctor: {
            type: String,
            required: true
        }
}, {timestamps: {createdAt: 'created', updatedAt: 'lastChange'} });

// Plugin para la paginación de documentos en la base de datos.
mongoose.plugin(mongoosePaginate)


// Exportación del modelo Patient
export const Patient = mongoose.model("patients", patientsSchema);

File: pdf-service/models/patient_model.py
from typing import Annotated, Literal
from pydantic import BaseModel, EmailStr, Field, field_validator, AfterValidator
from utils.dates_helper import valid_date

#Definición de tipos de texto con límites finitos
ShortStr = Annotated[str, Field(min_length=1, max_length=50)]     # Nombres, apellidos, afiliados
MediumStr = Annotated[str, Field(min_length=1, max_length=150)]   # Coberturas, direcciones
NoteStr = Annotated[str, Field(min_length=1, max_length=1000)]    # Descripciones individuales
DateYYMMDD = Annotated[str, AfterValidator(valid_date)]

class TreatmentItem(BaseModel):
    name: NoteStr
    dateStart: DateYYMMDD
    dateEnd: DateYYMMDD
    status: Literal["progress","pending", "finalized", "canceled"]


class ObservationItem(BaseModel):
    name: NoteStr
    date: DateYYMMDD
    status: Literal["progress","pending", "finalized", "canceled"]

class ToothStatusItem(BaseModel):
    tooth: int = Field(..., ge=11, le=85)  # Validación de rango para dientes (11-85)
    caries: dict[
        Literal["vestibular", "mesial", "oclusal", "distal", "lingual"],
        Literal["0", "1", "2"]
    ]
    corona: bool
    extracted: bool
    allcaries: bool
    incurable: bool
    malposition:bool 
    periodontal: bool
    inscrustration: bool

# Esquema del paciente
class PatientPDF(BaseModel):
    name: ShortStr
    lastName: ShortStr
    birth: DateYYMMDD

    # Opción A: Restringir a valores específicos conocidos
    typeDNI: Literal["DNI", "PASAPORTE", "LC", "LE", "CI"]
    dni: str = Field(..., min_length=5, max_length=10, pattern=r"^\d+$")
    
    # Opciones de sexo finitas (ajustar según la lógica de negocio)
    sex: Literal["male", "female", "anoher", "prefer_not_to_say"]
    
    address: MediumStr
    phone: str = Field(..., min_length=5, max_length=16, pattern=r"^\+?[0-9\s\-]+$")
    email: EmailStr = Field(..., max_length=254)
    
    medicalCoverage: MediumStr
    nAffiliate: ShortStr

    # 3. Validación de Listas (Límite de elementos + tipo de dato interno)
    treatments: list[TreatmentItem] = Field(..., max_length=200)
    observations: list[ObservationItem] = Field(..., max_length=100)
    dentalStatus: list[ToothStatusItem] = Field(..., max_length=50)

    